docs/spec/BeakerSpec.tex

```diff
@@ -0,0 +1,994 @@
+    \hline
+    \hline
+    \texttt{0x} & \texttt{00} & \texttt{00 00 00 00} & Proc Key & $\cdots$ & Procedure Heap \\
+    \texttt{0x} & \texttt{00} & \texttt{00 00 00 01} & \texttt{00 $\cdots$ 00} & \texttt{00 00 00} & \# of Procedures\\
+    \texttt{0x} & \texttt{00} & \texttt{00 00 00 01} & Key Index & \texttt{00 00 00} & Procedure List \\
+    \texttt{0x} & \texttt{00} & \texttt{00 00 00 02} & \texttt{00 $\cdots$ 00} & \texttt{00 00 00} & Kernel Address \\
+    \texttt{0x} & \texttt{00} & \texttt{00 00 00 03} & \texttt{00 $\cdots$ 00} & \texttt{00 00 00} & Current Procedure \\
+    \texttt{0x} & \texttt{00} & \texttt{00 00 00 04} & \texttt{00 $\cdots$ 00} & \texttt{00 00 00} & Entry Procedure \\
+    \hline
+  \end{tabularx}
+\end{table}
+
+\section{Procedures}\label{procedures}
+Data about procedure is store in 2 locations: the procedure table, and the
+procedure heap. The procedure table is simply a managed, enumerable, and
+iterable list of procedure keys (procedure key being something that identifies a
+particular procedure). The data associated with a procedure, such as its
+Ethereum address and its capabilities are stored on the procedure heap. Given a
+procedure key, the data associated with a procedure can be located using that
+key.
+
+\subsection{Procedure Key}
+Each procedure is defined by a procedure key. This key is a sequence of 24 8-bit
+bytes. This is treated as a sequence of 192~bits in all cases.
+
+The kernel itself treats the procedure keys as an opaque 192-bit value, however,
+the capability system and some functions apply prefixes to these key value. For
+example: capabilities that rely on procedure key values will often use a prefix
+value to define a range of keys. This allows one to defined a hierarchy of
+procedures, much like a directory structrue. See those capabilities for more
+detail.
+
+\subsection{Key Index}
+Each procedure key included in the kernel is associated with a key index, which
+identifies a procedure in the procedure table. This is not used anywhere except
+the procedure table itself. It is 1-based, that is the first value is 1, and the
+value 0 is a null value.
+
+As shown below, the maximum number of procedures is $2^{24} - 1 = 16,777,215$,
+therefore the maxium value of the key index is $16,777,215$, therefore the key
+index lies in the range $\left[1,1677215\right]$.
+
+For this reason the key index is specified as 24 bytes.
+
+\subsection{Procedure List}\label{procedure-table}
+The procedure list is simply an array of storage values. The first value is the
+length of this list, and the subsequent values are the procedure keys of the
+list.
+
+When a procedure is added to the kernel:
+
+\begin{enumerate}
+  \item The procedure data is added to the procedure heap (see
+  Section~\ref{procedure-heap}).
+  \item The procedure key is appended to the end of the array.
+  \item The length value (the first value) is incremented by one.
+\end{enumerate}
+
+When a procedure is deleted from the kernel:
```

Comment:
I would suggest to store the keyIndex in the Procedure Heap so that we don't have to iterate through the list. That way both insertion and deletion would be O(1)

Reply:
So for insertion:
1. The procedure data is added to the procedure heap
1. The procedure key is appended to the end of the array.
1. The length value (the first value) is incremented by one.
1. **The keyIndex value (in procedure metadata) is set to the length value**

On deletion:
1. If the procedure key is the same as the Entry Procedure Key, abort and
  throw an error.
1. If the procedure key does not exist in the list, abort and throw an
  error.
1. The length value is decremented by one.
1. **Using the keyIndex lookup if the procedure being deleted is not the last in the list.** If the last in
  the list is copied to overwrite the key being deleted. This also accounts for
  the case of an empty list.

Reply:
This has been added.